src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java

```diff
@@ -1118,6 +1121,84 @@ public void insertCode(String code, boolean blockMode)
    {
       widget_.getEditor().insert(StringUtil.normalizeNewLines(code));
    }
+   
+   public void applyCodeChanges(JsdiffChange[] changes)
+   {
+      // alias apis
+      AceEditorNative editor = widget_.getEditor();
+      EditSession session = editor.getSession();
+      Selection selection = session.getSelection();
+      AceCommandManager commandManager = editor.getCommandManager();
+      
+      // function to advance the selection
+      Consumer<Integer> advanceSelection = (Integer charsLeft) -> {
+         
+         // start from current line/row
+         Position startPos = selection.getCursor();
+         
+         // iterate through rows until we've consumed all the chars
+         int row;
+         int col = startPos.getColumn();
+         for (row = startPos.getRow(); row < session.getLength(); row++) {
+            
+            // how many chars left in the current column?
+            String line = session.getLine(row);
+            int charsLeftInLine = line.length() - col;
+            
+            // is the number of chars we still need to consume lte
+            // the number of charsLeft?
+            if (charsLeft <= charsLeftInLine) 
+            {
+               col = col + charsLeft;
+               break;
+            }
+            else
+            {
+               charsLeft -= (charsLeftInLine+1); // add 1 for newline
+               col = 0;
+            }
+         }
```

Comment:
The logic looks correct to me. I had to focus hard to make sure that a) `charsLeft` can never become less than 0, and b) that the row will never be advanced accidentally.

This suggestion is (supposed to be) equivalent but a little easier for my brain. I changed the for-loop to while, incrementing `row` manually; and moved the newline "+1" from the end of the loop body to the beginning, which I find makes the if/else easier to reason about.

```java
         int row = startPos.getRow();
         int col = startPos.getColumn();
         while (row < session.getLength()) {

            // how many chars left in the current column?
            String line = session.getLine(row);
            // the +1 is for newline
            int charsLeftInLine = line.length() + 1 - col;

            // is the number of chars we still need to consume lte
            // the number of charsLeft?
            if (charsLeft < charsLeftInLine) 
            {
               col += charsLeft;
               break;
            }
            else
            {
               charsLeft -= charsLeftInLine;
               col = 0;
               row++;
            }
         }
```

Reply:
Yes, your version reach much nicer. Made that change. Thanks very much for the review, this is not a bit of code I'd ever want to get wrong!